SNIPS/snip-20.md

```diff
@@ -0,0 +1,120 @@
+
+We propose that full nodes indiscriminantly propagate paymaster transactions without looking into the `paymaster` field; only sequencers will bar entry into their respective mempools.
+
+Now the flow:
+
+1. Syntactic checks (transaction is not a cat gif).
+2. User submitted rate should be at least client's locally configured rate. $$\mathtt{min\_ STRK/TOKEN\_ rate}\leq r$$
+3. Balance checks:
+    * User token balance should cover their bid given their specified max exchange rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$
+
+#### Block builder
+
+A paymaster contract should be used only by those who trust its code. Such users know the paymaster's exchange policies, and therefore know which rates they should submit to pass, and which rates are too frugal and may cause `__validate_paymaster__` to fail. Hence, we consider it justified to "blame" users for failed `__validate_paymaster__` and to consequently charge them. Now arises a problematic scenario: the user may hold zero STRK and the paymaster rejected the request to cover its transaction fee. In this particular case we allow the sequencer to charge the transaction fee directly in the user's designated token.
+
+In the naive example above, `__validate_paymaster__` will fail during block building if and only if $$\mathtt{min\_ STRK/TOKEN\_ rate} \leq r\leq 1.05 \times \mathtt{oracle\_ rate},$$ meaning the user's rate was high enough to enter the client's mempool but too low for the paymaster.
+
+Now the flow:
+
+1. Balance checks:
+    * User token balance should cover their bid given their specified max rate $$\mathtt{balance(user)}\geq \mathtt{max\_ amount}\times r(\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+    * Paymaster STRK balance should cover user bid $$\mathtt{balance(paymaster)}\geq \mathtt{max\_ amount}\times (\mathtt{base\_ price(current\_ block)}+\mathtt{tip}).$$ If success, CONTINUE; else, REJECT.
+2. Run account `__validate__`. If success CONTINUE, else REJECT.
+3. Run paymaster `__validate_paymaster__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+4. Run account `__execute__`. If success, CONTINUE; else, REVERT and skip to fee invocations below.
+5. Fee invocations:
+    * If failure was in `__validate_paymaster__`, user pays sequencer in designated fee token:
+        * Burn $\mathtt{used\_ amount}\times r\cdot\mathtt{base\_ price}$.
+        * Transfer $\mathtt{used\_ amount}\times r\cdot\mathtt{tip}$.
+    * If failure was later:
+        1. Paymaster pays transaction fee to sequencer:
```

Comment:
What happens if the paymaster does not have the required balance?

Reply:
The paymaster had the required balance at the beginning, and it seems reasonable to expect this to hold at the end of the tx if the paymaster is honest. So a solution is simply to revert the tx, and do the fee invocations on the used amount (from user and from PM, which were both holding enough).

Reply:
Don't you think the most reasonable paymaster implementation is to do the token conventions in the transactions?

Do you expect the paymasters to hold a large STARK balance and do periodic token conversions?